### PR TITLE
Publish solutions for lab 12

### PR DESCRIPTION
**Pull request by sketch873**

### CODE REVIEW

File: labs/lab12/sol/lin/1-mini/mini.c [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+#endif
+
+#ifdef TODO3
+		/* TODO3: implement mount source target fs_type
+		 * e.g: mount /dev/sda1 /mnt/disk2 ext3
+		 */
+		if (strncmp(cmd, "mount", 5) == 0) {
+			arg1 = strtok(NULL, delim); /* source */
+			arg2 = strtok(NULL, delim); /* target */
+			arg3 = strtok(NULL, delim);/* fs_type (e.g: ext2) */
```
Comment (teodutu) [inline labs/lab12/sol/lin/1-mini/mini.c:124]:
```suggestion
			arg3 = strtok(NULL, delim); /* fs_type (e.g: ext2) */
```

File: labs/lab12/sol/lin/1-mini/mini.c [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+char *prompt = "so-lab12";
+
+#define TODO2
+#define TODO3
+#define TODO4
+#define TODO5
+#define TODO6
+#define TODO7
+
+/* Recursively list all dirs under 'dirname' path
```
Comment (teodutu) [inline labs/lab12/sol/lin/1-mini/mini.c:32]:
```suggestion
/**
 * Recursively list all dirs under 'dirname' path
```

File: labs/lab12/sol/lin/1-mini/mini.c [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+		cmd = strtok(line, delim);
+		if (!cmd)
+			continue;
+#ifdef DEBUG
+		printf("Executing command: %s\n", cmd);
+#endif
+		if (strncmp(cmd, "quit", 4) == 0)
+			break;
+#ifdef TODO2
+		/* TODO2: implement list <device_node>
```
Comment (teodutu) [inline labs/lab12/sol/lin/1-mini/mini.c]:
```suggestion
		/**
		 * TODO2: implement list <device_node>
```

File: labs/lab12/sol/lin/1-mini/mini.c [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+			if (S_ISBLK(s_buf.st_mode))
+				type = 'B';
+
+			printf("%s: <%c> %d:%d\n", arg1, type,
+				major(s_buf.st_rdev), minor(s_buf.st_rdev));
+		}
+#endif
+
+#ifdef TODO3
+		/* TODO3: implement mount source target fs_type
```
Comment (teodutu) [inline labs/lab12/sol/lin/1-mini/mini.c]:
```suggestion
		/**
		 * TODO3: implement mount source target fs_type
```

File: labs/lab12/sol/lin/1-mini/mini.c [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+			/* TODO3: implement umount */
+			arg1 = strtok(NULL, delim); /* target */
+
+			ret = umount(arg1);
+			DIE(ret < 0, "umount failed");
+		}
+#endif
+
+#ifdef TODO4
+		/* TODO4: implement symlink oldpath newpath
```
Comment (teodutu) [inline labs/lab12/sol/lin/1-mini/mini.c]:
```suggestion
		/**
		 * TODO4: implement symlink oldpath newpath
```

File: labs/lab12/sol/lin/1-mini/mini.c [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+			/* TODO4: implement unlink */
+			arg1 = strtok(NULL, delim); /* pathname */
+
+			ret = unlink(arg1);
+			DIE(ret < 0, "unlinking failed");
+		}
+#endif
+
+#ifdef TODO5
+		/* TODO5: implement mkdir pathname mode
```
Comment (teodutu) [inline labs/lab12/sol/lin/1-mini/mini.c]:
```suggestion
		/**
		 * TODO5: implement mkdir pathname mode
```

File: labs/lab12/sol/lin/1-mini/mini.c [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+			/* TODO5: implement rmdir pathname */
+			arg1 = strtok(NULL, delim); /* pathname */
+
+			ret = rmdir(arg1);
+			DIE(ret < 0, "rmdir failed");
+		}
+#endif
+
+#ifdef TODO6
+		/* TODO6: implement ls dirname
```
Comment (teodutu) [inline labs/lab12/sol/lin/1-mini/mini.c]:
```suggestion
		/**
		 * TODO6: implement ls dirname
```

File: labs/lab12/sol/lin/1-mini/mini.c [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+		if (strncmp(cmd, "ls", 2) == 0) {
+			/* recursively print files starting with arg1 */
+			arg1 = strtok(NULL, delim);
+			list_dir(arg1);
+		}
+#endif
+
+#ifdef TODO7
+		if (strncmp(cmd, "chdir", 5) == 0) {
+			/* TODO7: implement chdir <dir>
```
Comment (teodutu) [inline labs/lab12/sol/lin/1-mini/mini.c]:
```suggestion
			/**
			 * TODO7: implement chdir <dir>
```

File: labs/lab12/sol/lin/1-mini/mini.c [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+			 * e.g: chdir bar
+			 */
+			arg1 = strtok(NULL, delim); /* pathname */
+			ret = chdir(arg1);
+			DIE(ret < 0, "chdir failed");
+		}
+
+		if (strncmp(cmd, "pwd", 3) == 0) {
+			/* TODO7: implement pwdir
```
Comment (teodutu) [inline labs/lab12/sol/lin/1-mini/mini.c]:
```suggestion
			/**
			 * TODO7: implement pwdir
```

File: labs/lab12/sol/lin/1-mini/utils.h [hunk 10]
```diff
@@ -0,0 +1,23 @@
+/**
+ *
```
Comment (teodutu) [inline labs/lab12/sol/lin/1-mini/utils.h:4]:
```suggestion
 * SO
 * Linux
```

**Review by teodutu — CHANGES_REQUESTED** [review]
On top of the inline suggestions I made, I think you should convert `lab12/sol/lin/1-mini/README` to Markdown format.

**Review by teodutu — DISMISSED** [review]

File: labs/lab12/sol/lin/1-mini/README.md [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+brw-rw---T 1 root disk 8, 1 May 14 19:50 /dev/sda1
+b - block device, 8 - major, 1 - minor
+```
+
+## Exerciţiul 3
+* trebuie să includeţi `<sys/mount.h>`
+
+## Exerciţiul 6
+* man opendir
+* trebuie incluse `<sys/types.h>` şi `<dirent.h>`
```
Comment (andreia-oca) [inline labs/lab12/sol/lin/1-mini/README.md]: # Additional Hints

## Task 1
* `ls -l /dev` will list all node devices. 
* Check the snippet below:

```
# b - block device, 8 - major, 1 - minor
brw-rw---T 1 root disk 8, 1 May 14 19:50 /dev/sda1
```

## Task 3
* Include `<sys/mount.h>`.

## Task 6
* `man opendir`
* Include `<sys/types.h>` and `<dirent.h>`.

Comment (andreia-oca) [reply labs/lab12/sol/lin/1-mini/README.md]: I would translate this whole README in English.

**Review by andreia-oca — DISMISSED** [review]

**Review by andreia-oca — APPROVED** [review]

**Review by teodutu — APPROVED** [review]

**Pull request merged**